pkg/kubefed2/unjoin.go

```diff
@@ -0,0 +1,409 @@
+		glog.V(2).Infof("Failed to get federation clientset: %v", err)
+		return err
+	}
+
+	glog.V(2).Infof("Performing preflight checks for unjoin cluster: %s.", unjoiningClusterName)
+	err = performPreflightChecksForUnjoin(clusterClientset, unjoiningClusterName, host, federationNamespace)
+	if err != nil {
+		return err
+	}
+
+	if removeFromRegistry {
+		err = removeFromClusterRegistry(hostConfig, clusterConfig.Host, unjoiningClusterName, dryRun)
+		if err != nil {
+			return err
+		}
+	}
+
+	glog.V(2).Infof("Deleting federated cluster resource from namespace: %s for unjoin cluster: %s",
+		federationNamespace, unjoiningClusterName)
+
+	_, err = deleteFederatedClusterAndSecret(hostClientset, fedClientset, federationNamespace,
+		unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Failed to delete federated cluster resource from namespace: %s for unjoin cluster: %s due to: %v",
+			federationNamespace, unjoiningClusterName, err)
+		return err
+	}
+
+	glog.V(2).Infof("Deleted federated cluster resource from namespace: %s for unjoin cluster: %s",
+		federationNamespace, unjoiningClusterName)
+
+	err = deleteRBACResources(hostClientset, clusterClientset,
+		federationNamespace, unjoiningClusterName, host, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Could not delete cluster RBAC resources: %v", err)
+		return err
+	}
+
+	if unjoiningClusterContext == host {
+		glog.V(2).Infof("Host cluster: %s and unjoin cluster: %s are same, no need to delete federation namespace: %s from unjoin cluster.",
+			host, unjoiningClusterContext, federationNamespace)
+	} else {
+		glog.V(2).Infof("Deleting federation namespace: %s from unjoin cluster: %s",
+			federationNamespace, unjoiningClusterName)
+
+		err = deleteFedNSFromUnjoinCluster(clusterClientset,
+			federationNamespace, unjoiningClusterName, host, dryRun)
+		if err != nil {
+			glog.V(2).Infof("Could not delete cluster RBAC resources: %v", err)
+			return err
+		}
+
+		glog.V(2).Infof("Deleted federation namespace: %s from unjoin cluster: %s",
+			federationNamespace, unjoiningClusterName)
+	}
+
+	return nil
+}
+
+// performPreflightChecksForUnjoin checks that the host and unjoining clusters are in
+// a consistent state.
+func performPreflightChecksForUnjoin(clusterClientset client.Interface, name, host,
+	federationNamespace string) error {
+	// Make sure there is a existing service account in the unjoining cluster.
+	saName := util.ClusterServiceAccountName(name, host)
+	sa, err := clusterClientset.CoreV1().ServiceAccounts(federationNamespace).Get(saName,
+		metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		return fmt.Errorf("service account does not exist in unjoining cluster")
+	} else if err != nil {
+		return err
+	} else if sa != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// removeFromClusterRegistry handles removing the cluster from the cluster registry and
+// reports progress.
+func removeFromClusterRegistry(hostConfig *rest.Config, host, unjoiningClusterName string,
```

Comment:
(No action required) Addition/removal from cluster registry might be a good thing for the cluster registry to be responsible for, so that vendoring repos can reuse rather than reimplement this basic functionality.

Reply:
+1